tls/s2n_client_hello.c

```diff
@@ -56,6 +56,34 @@ static uint32_t min_size(struct s2n_blob *blob, uint32_t max_length) {
     return blob->size < max_length ? blob->size : max_length;
 }
 
+static S2N_RESULT s2n_generate_client_session_id(struct s2n_connection *conn)
+{
+    ENSURE_REF(conn);
+
+    /* Session id already generated - no-op */
+    if (conn->session_id_len) {
+        return S2N_RESULT_OK;
+    }
+
+    /* Only generate the session id for pre-TLS1.3 if using tickets */
+    if (conn->client_protocol_version < S2N_TLS13 && !conn->config->use_tickets) {
+        return S2N_RESULT_OK;
+    }
+
+    /* Generate the session id for TLS1.3 if in middlebox compatibility mode.
+     * For now, we default to middlebox compatibility mode unless using QUIC. */
```

Comment:
Is there a way to not use middlebox compatibility mode?

Reply:
No. Before this change, middlebox compatibility mode was on by default, we just weren't _quite_ doing it right. We were sending the CCS messages, but not setting this field.

After this change, middlebox compatibility mode is still on by default, but done correctly.